src/channel/annotations.rs

```diff
@@ -0,0 +1,97 @@
+use std::fmt;
+
+#[derive(Default)]
+pub struct Annotations {
+    pub scope: Vec<String>,
+    pub annotations: Vec<String>,
+    pub annotations_enabled: bool,
+    pub extra_annotations: Vec<String>,
+    pub extra_annotations_enabled: bool,
+    pub prover_to_verifier_bytes: usize,
+    pub expected_annotations: Option<Vec<String>>,
+    pub annotation_prefix: String,
+}
+
+impl Annotations {
+    pub fn update_annotation_prefix(&mut self) {
+        self.annotation_prefix = self
+            .scope
+            .iter()
+            .fold(String::new(), |acc, s| acc + "/" + s)
+            + ": ";
+    }
+
+    pub fn add_annotation(&mut self, annotation: String) {
+        assert!(
+            self.annotations_enabled,
+            "Cannot add annotation when annotations are disabled."
+        );
+        if let Some(expected_annotations) = &self.expected_annotations {
+            let idx = self.annotations.len();
+            assert!(
+                idx < expected_annotations.len(),
+                "Expected annotations is too short."
+            );
+            let expected_annotation = &expected_annotations[idx];
+            assert!(
+                expected_annotation == &annotation,
+                "Annotation mismatch. Expected annotation: '{}'. Found: '{}'",
+                expected_annotation,
+                annotation
+            );
+        }
+        self.annotations.push(annotation);
+    }
+
+    pub fn add_extra_annotation(&mut self, annotation: String) {
+        assert!(
+            self.extra_annotations_enabled,
+            "Cannot add extra annotation when extra annotations are disabled."
+        );
+        // XXX : why add prefix here, while add_annotation doesn't? XD
+        self.extra_annotations.push(format!("{}{}\n", self.annotation_prefix, annotation));
+    }
+
+    pub fn annotate_prover_to_verifier(&mut self, annotation: String, n_bytes: usize) {
```

Comment:
Consider introducing an annotation type rather than having everything as String, ala.

```
enum Annotation {
    ProverToVerifier(n_bytes),
    VerifierToProver(..)
}
```

In general: sum types are one of the things which is much less painful in Rust than C++.

Reply:
This also prevents the need for e.g separate:

`annotate_prover_to_verifier`

And

`annotate_verifier_to_prover`

Methods